In [60]:
#Importing libraries
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt

In [61]:
#Importing the dataset
df = pd.read_csv("50_Startups.csv")
df.head()

R&D Spend  Administration  Marketing Spend       State     Profit
0  165349.20       136897.80        471784.10    New York  192261.83
1  162597.70       151377.59        443898.53  California  191792.06
2  153441.51       101145.55        407934.54     Florida  191050.39
3  144372.41       118671.85        383199.62    New York  182901.99
4  142107.34        91391.77        366168.42     Florida  166187.94

In [62]:
#shape of data
df.shape

(50, 5)

In [63]:
# divided dataset into X and y 
# OR
#X = df.iloc[:,:-1]
#y = df.iloc[:,4]
X = df.drop(['Profit'],axis=1)
y = df['Profit']

In [64]:
#Count State 
df['State'].value_counts()

California    17
New York      17
Florida       16
Name: State, dtype: int64

In [65]:
#Data preprocessing
# Convert the  categorical columns into numerical columns
states = pd.get_dummies(X['State'],drop_first=True)
states.head()

Florida  New York
0        0         1
1        0         0
2        1         0
3        0         1
4        1         0

In [68]:
# Drop the state column because I created already State dummies colums
X = X.drop('State',axis=1)
X.head()

R&D Spend  Administration  Marketing Spend
0  165349.20       136897.80        471784.10
1  162597.70       151377.59        443898.53
2  153441.51       101145.55        407934.54
3  144372.41       118671.85        383199.62
4  142107.34        91391.77        366168.42

In [69]:
#concat the dummy variables
X = pd.concat([X,states],axis=1)
X.head()

R&D Spend  Administration  Marketing Spend  Florida  New York
0  165349.20       136897.80        471784.10        0         1
1  162597.70       151377.59        443898.53        0         0
2  153441.51       101145.55        407934.54        1         0
3  144372.41       118671.85        383199.62        0         1
4  142107.34        91391.77        366168.42        1         0

In [72]:
# Equation y = B0+B1.x1+B2.x2.....+Bn.Xn
#Splitting the dataset into the training and test set
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=0)
print("X_train split data size ",X_train.shape)
print("X_test  split data size ",X_test.shape)
print("y_train split data size ",y_train.shape)
print("y_test  split data size ",y_test.shape)

X_train split data size  (40, 5)
X_test  split data size  (10, 5)
y_train split data size  (40,)
y_test  split data size  (10,)


In [73]:
# fitting multiple linear regression to the training set
from sklearn.linear_model import LinearRegression
multipleRegression = LinearRegression()
multipleRegression.fit(X_train,y_train)

LinearRegression()

In [76]:
#predict the test set result
y_pred = multipleRegression.predict(X_test)


In [78]:
#compare y_test and y_predict value
print(y_test)
print("\n--------------Compare y_test and y_predict-------------- \n")
print(y_pred)

28    103282.38
11    144259.40
10    146121.95
41     77798.83
2     191050.39
27    105008.31
38     81229.06
31     97483.56
22    110352.25
4     166187.94
Name: Profit, dtype: float64

--------------Compare y_test and y_predict-------------- 

[103015.20159796 132582.27760816 132447.73845174  71976.09851258
 178537.48221055 116161.24230165  67851.69209676  98791.73374687
 113969.43533012 167921.0656955 ]


In [80]:
# R square formula
# R^2  = 1 - SSres/SSmean
# SSres = 1/n summation i=1 to n (y-ypred)^2
# SSmean = 1/n summation i=1 to n (y-ymean)^2
# SSmean always greater than SSres (SSmean > SSres) if our model is very very good
# range 0 to 1
from sklearn.metrics import r2_score
score = r2_score(y_test,y_pred)
score

0.9347068473282423

Its is a good model r2 value is 0.93 that near to 1 .I'm using this r2 score to measure how my model has done